### PR TITLE
747 pandas df function

### PR DESCRIPTION
**Pull request by CharlesRendle**
PR addressing issue #747

Adds an inspect API function to load a pandas dataframe from the provided `csv_url`
The dataframe loads each column with the correct data type corresponding to the 'column_definition` for obs columns and attribute literal columns. Other columns are assigned as strings.

Two unit tests added to demonstrate this functionality (two test cases also added).

Also adds `get_primary_csv_url()` functions to both code_list_inspector and data_cube_inspector classes which retrieves the csv_url for the primary CSV defined in the CSV-W. Included two unit tests to demonstrate this function (one for each class)

### CODE REVIEW

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py [hunk 1]
```diff
@@ -207,6 +217,50 @@ def get_code_lists_and_cols(self, csv_url: str) -> CodelistsResult:
 
         return self._codelists_and_cols.get(csv_url, CodelistsResult([], 0))
 
+    def get_dataframe(self, csv_url: str) -> Tuple[pd.DataFrame, List[ValidationError]]:
+        """
+        Get the pandas dataframe for the csv url of the cube wishing to be loaded.
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_inspector.py:222]: It's be good to explain what the list of validation errors we return are about in this comment too.

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py [hunk 2]
```diff
@@ -207,6 +217,50 @@ def get_code_lists_and_cols(self, csv_url: str) -> CodelistsResult:
...
+                        col.column_definition.title
+                    ] = ACCEPTED_DATATYPE_MAPPING[col_data_type]
+                else:
+                    raise ValueError(
+                        f"Unhandled data type '{col.column_definition.data_type}'."
+                    )
+            else:
+                dict_of_types[col.column_definition.title] = "string"
+
+        absolute_csv_url = Path(
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_inspector.py:255]: Could you refactor this into a utils function which takes a URI and returns a file path? Ideally in the comment you could explain that the function works on both windows and lnux/unix.

There are a couple of other locations in the code base which do this same kind of thing. Could you tidy them up to use this new function too?

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py [hunk 3]
```diff
@@ -207,6 +217,50 @@ def get_code_lists_and_cols(self, csv_url: str) -> CodelistsResult:
 
         return self._codelists_and_cols.get(csv_url, CodelistsResult([], 0))
 
+    def get_dataframe(self, csv_url: str) -> Tuple[pd.DataFrame, List[ValidationError]]:
+        """
+        Get the pandas dataframe for the csv url of the cube wishing to be loaded.
+        """
+        cols = self.get_column_component_info(csv_url)
+        dict_of_types = {}
+        for col in cols:
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_inspector.py:226]: This function seems to be getting a bit big now. Could you simplify it a bit by refactoring some of the code into a separate function? Perhaps the stuff which generates the `dict_of_types` makes sense to put into a separate function?

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py [hunk 4]
```diff
@@ -207,6 +217,50 @@ def get_code_lists_and_cols(self, csv_url: str) -> CodelistsResult:
...
+                dict_of_types[col.column_definition.title] = "string"
+
+        absolute_csv_url = Path(
+            os.path.normpath(
+                urljoin(self.csvw_inspector.csvw_json_path.as_uri(), csv_url)
+            )
+            .removeprefix("file:\\")
+            .removeprefix("file:")
+        )
+        return read_csv(Path(absolute_csv_url), dtype=dict_of_types)
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_inspector.py:262]: `absolute_csv_url` is already a Path at this point.

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py [hunk 5]
```diff
@@ -207,6 +217,50 @@ def get_code_lists_and_cols(self, csv_url: str) -> CodelistsResult:
...
+                col.column_type == EndUserColumnType.Observations
+                or is_attribute_literal
+            ):
+                if col.column_definition.data_type is not None:
+                    col_data_type = col.column_definition.data_type.removeprefix(
+                        _XSD_BASE_URI
+                    )
+                else:
+                    raise ValueError(
+                        "Expected a defined datatype but got 'None' instead."
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_inspector.py:242]: Can you tell us which column it is that's the problem in this message? It'll help debugging when this inevitably happens.

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py [hunk 6]
```diff
@@ -207,6 +217,50 @@ def get_code_lists_and_cols(self, csv_url: str) -> CodelistsResult:
...
+            is_attribute_literal = (
+                col.column_type == EndUserColumnType.Attribute
+                and col.column_definition.value_url is None
+            )
+
+            if (
+                col.column_type == EndUserColumnType.Observations
+                or is_attribute_literal
+            ):
+                if col.column_definition.data_type is not None:
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_inspector.py:236]: I think this code would read better if it were like this:

```python
if col.column_definition.data_type is None:
   raise ValueError(...)

col_data_type = col.column_definition.data_type.removeprefix(...)
```

That way the assignment of col_data_type isn't hidden away in an if-block.

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py [hunk 7]
```diff
@@ -207,6 +217,50 @@ def get_code_lists_and_cols(self, csv_url: str) -> CodelistsResult:
...
+                    col_data_type = col.column_definition.data_type.removeprefix(
+                        _XSD_BASE_URI
+                    )
+                else:
+                    raise ValueError(
+                        "Expected a defined datatype but got 'None' instead."
+                    )
+                if col_data_type in ACCEPTED_DATATYPE_MAPPING.keys():
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_inspector.py:244]: You can just say `col_data_type in ACCEPTED_DATATYPE_MAPPING`, without the `.keys()` bit here.

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py [hunk 8]
```diff
@@ -207,6 +217,50 @@ def get_code_lists_and_cols(self, csv_url: str) -> CodelistsResult:
...
+                    dict_of_types[
+                        col.column_definition.title
+                    ] = ACCEPTED_DATATYPE_MAPPING[col_data_type]
+                else:
+                    raise ValueError(
+                        f"Unhandled data type '{col.column_definition.data_type}'."
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_inspector.py:250]: Again, can you mention the column title in this error message?

File: tests/unit/utils/sparqlhandler/test_data_cube_inspector.py [hunk 9]
```diff
@@ -902,3 +906,73 @@ def test_get_columns_for_component_attribute_pivoted():
...
+        / "testing-converting-a-pivoted-csvw-to-pandas-dataframe.csv-metadata.json"
+    )
+    data_cube_inspector = get_data_cube_inspector(csvw_metadata_json_path)
+    csv_url = data_cube_inspector.get_primary_csv_url()
+
+    dataframe, validation_errors = data_cube_inspector.get_dataframe(csv_url)
+
+    assert isinstance(dataframe, pd.DataFrame)
+    assert dataframe["Dim1"].dtype == "string"
+    assert dataframe["Obs1"].dtype == "string"
```
Comment (robons) [inline tests/unit/utils/sparqlhandler/test_data_cube_inspector.py:975]: Is this the one that has a obs val column which has the time data type?

**Review by robons — COMMENTED** [review]
Looks pretty close to me, just a bit of refactoring if you could.

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py (hunk 1 above)
Comment (CharlesRendle) [reply src/csvcubed/utils/sparql_handler/data_cube_inspector.py:222]: Added further comment.

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py (hunk 3 above)
Comment (CharlesRendle) [reply src/csvcubed/utils/sparql_handler/data_cube_inspector.py]: refactored the `dict_of_types` bit into a private function.

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py (hunk 4 above)
Comment (CharlesRendle) [reply src/csvcubed/utils/sparql_handler/data_cube_inspector.py]: removed

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py (hunk 5 above)
Comment (CharlesRendle) [reply src/csvcubed/utils/sparql_handler/data_cube_inspector.py]: added the column title to the message

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py (hunk 6 above)
Comment (CharlesRendle) [reply src/csvcubed/utils/sparql_handler/data_cube_inspector.py]: rearranged logic blocks

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py (hunk 7 above)
Comment (CharlesRendle) [reply src/csvcubed/utils/sparql_handler/data_cube_inspector.py]: removed `.keys()`

File: src/csvcubed/utils/sparql_handler/data_cube_inspector.py (hunk 8 above)
Comment (CharlesRendle) [reply src/csvcubed/utils/sparql_handler/data_cube_inspector.py]: added the column title to the message

File: tests/unit/utils/sparqlhandler/test_data_cube_inspector.py (hunk 9 above)
Comment (CharlesRendle) [reply tests/unit/utils/sparqlhandler/test_data_cube_inspector.py:977]: Yeah it was the `time` column. Have added a comment to explain why we expect "string"

File: tests/unit/utils/test_uri.py [hunk 10]
```diff
@@ -53,5 +57,24 @@ def test_ensure_all_look_like_uri():
...
+def test_get_absolute_file_path():
+    """
+    Testing the URI util function `get_absolute_file_path()` returns a normalised
+    path from a string URI which can be used to locate resources on windows as
+    well as unix/linux operating systems.
+    """
+    csv_url = "file:///workspaces/csvcubed/tests/test-cases/cli/inspect/inspector-load-dataframe/pivoted-shape/pivoted-shape-out/testing-converting-a-pivoted-csvw-to-pandas-dataframe.csv"
+    absolute_csv_url = get_absolute_file_path(csv_url)
+    assert isinstance(absolute_csv_url, Path)
+    if os.name == "nt":
```
Comment (robons) [inline tests/unit/utils/test_uri.py:69]: I think you should have different `csv_urls` for each OS. The windows one should probably start with`file:/C:/...`.

**Review by robons — APPROVED** [review]
Approved, but could you address the two small comments I've made?

**Pull request merged**